TFT/src/User/API/Printing.c

```diff
@@ -6,6 +6,29 @@ PRINTING infoPrinting;
 
 static bool update_waiting = false;
 
+static float filament_used;
+static float last_E_pos;
+
+void resetFilamentUsed(void)
+{
+  filament_used = 0;
+  last_E_pos = ((infoFile.source == BOARD_SD) ? coordinateGetAxisActual(E_AXIS) : coordinateGetAxisTarget(E_AXIS));
+}
+
+void updateFilamentUsed(void)
+{
+  float E_pos = ((infoFile.source == BOARD_SD) ? coordinateGetAxisActual(E_AXIS) : coordinateGetAxisTarget(E_AXIS));
+  if ((E_pos + 20) < last_E_pos) //Check whether E position reset. If retract more than 20mm, false filament used values would be calculated.
+  {
+    filament_used = filament_used + E_pos;
+    last_E_pos = E_pos;
+  } else if (E_pos > last_E_pos)
```

Comment:
I suspect that making the calculation at the end of the print would produce wrong results when there are G92 resets at the mid of the prints.

Lets assume print started at E=0 and Last_E_Pos is 0. At E=100 a G92 reset ocuurs and since the filament used and last_E_Pos is not updated till that point it will stay at 0. If we make such change the result would only contain the filament used after the last G92 reset.

Please advice if I am missing something.

Reply:
Yup, you're right, I was thinking only about the `filament_used = filament_used + (E_pos - last_E_pos);` part of that "else".

I was thinking about something like this:
`void updateFilamentUsed(void)`
`{`
`__float E_pos = ((infoFile.source == BOARD_SD) ? coordinateGetAxisActual(E_AXIS) : coordinateGetAxisTarget(E_AXIS));`
`__if ((E_pos + 20) < last_E_pos) //Check whether E position reset. If retract more than 20mm, false filament used values would be calculated.`
`__{`
`____filament_used = filament_used + E_pos;`
`__} `
`__last_E_pos = E_pos;`
`__if !isPrinting()`
`______filament_used = filament_used + (E_pos - last_E_pos);`
`}`